api/odilo.py

```diff
@@ -0,0 +1,495 @@
+                   {"key": ExternalIntegration.PASSWORD, "label": _("Client Secret")},
+               ] + BaseCirculationAPI.SETTINGS
+
+    SET_DELIVERY_MECHANISM_AT = BaseCirculationAPI.FULFILL_STEP
+
+    # maps a 2-tuple (media_type, drm_mechanism) to the internal string used in Odilo API to describe that setup.
+    delivery_mechanism_to_internal_format = {
+        v: k for k, v in OdiloRepresentationExtractor.format_data_for_odilo_format.iteritems()
+        }
+
+    error_to_exception = {
+        "TitleNotCheckedOut": NoActiveLoan,
+        "patronNotFound": PatronNotFoundOnRemote,
+        "ERROR_DATA_NOT_FOUND": NotFoundOnRemote,
+        "LOAN_ALREADY_RESERVED": AlreadyOnHold,
+        "CHECKOUT_NOT_FOUND": NotCheckedOut,
+    }
+
+    def __init__(self, _db, collection):
+        super(OdiloAPI, self).__init__(_db, collection)
+        self.odilo_bibliographic_coverage_provider = (
+            OdiloBibliographicCoverageProvider(
+                collection, api_class=self
+            )
+        )
+
+    def patron_request(self, patron, pin, url, extra_headers={}, data=None, exception_on_401=False, method=None):
+        """Make an HTTP request on behalf of a patron.
+
+        The results are never cached.
+        """
+        headers = dict(Authorization="Bearer %s" % self.token)
+        headers['Content-Type'] = 'application/json'
+        headers.update(extra_headers)
+
+        if method and method.lower() in ('get', 'post', 'put', 'delete'):
+            method = method.lower()
+        else:
+            if data:
+                method = 'post'
+            else:
+                method = 'get'
+
+        response = HTTP.request_with_timeout(method, self.library_api_base_url + url, headers=headers, data=data,
+                                             timeout=60)
+        if response.status_code == 401:
+            if exception_on_401:
+                # This is our second try. Give up.
+                raise Exception("Something's wrong with the patron OAuth Bearer Token!")
+            else:
+                # Refresh the token and try again.
+                self.check_creds(True)
+                return self.patron_request(patron, pin, url, extra_headers, data, True)
+        else:
+            return response
+
+    def get_patron_credential(self, patron, pin):
+        """Create an OAuth token for the given patron."""
+
+        def refresh(credential):
+            return self.get_patron_access_token(credential, patron, pin)
+
+        return Credential.lookup(self._db, DataSource.ODILO, "OAuth Token", patron, refresh)
+
+    def get_patron_access_token(self, credential, patron, pin):
+        """Request an OAuth bearer token that allows us to act on
+        behalf of a specific patron.
+        """
+
+        self.client_key = patron
+        self.client_secret = pin
+        message = self.refresh_creds(credential)
+
+        if 'OK' == message:
```

Comment:
See my core comment https://github.com/NYPL-Simplified/server_core/pull/695/files#r150567737

Reply:
Yes, I've already changed it.